src/fiber/fiber.ml

```diff
@@ -480,16 +479,19 @@ module Throttle = struct
           f ())
 end
 
-let apply_and_set_result f =
-  let result = ref None in
-  EC.apply f () (fun x -> result := Some x);
-  result
-
-let run t = EC.new_run (fun () -> !(apply_and_set_result (fun () -> t)))
+type fill = Fill : 'a Ivar.t * 'a -> fill
 
-let run2 a b =
+let run t ~iter =
   EC.new_run (fun () ->
-      EC.add_refs 1;
-      let result_a = apply_and_set_result a in
-      let result_b = apply_and_set_result b in
-      (!result_a, !result_b))
+      let result = ref None in
+      EC.apply (fun () -> t) () (fun x -> result := Some x);
+      let rec loop () =
+        match !result with
+        | Some res -> res
+        | None ->
+          let ivars = iter () in
+          if List.is_empty ivars then Code_error.raise "Fiber.run got stuck" [];
+          List.iter ivars ~f:(fun (Fill (ivar, v)) -> Ivar.fill ivar v ignore);
```

Comment:
One thing that's possibly undesirable is that we're filling more ivars than necessary to fill `result`. Is that intended

Reply:
Yeah, I suppose that's not ideal. We could also have `iter` return a single ivar to fill. That would avoid raising in case of the empty list as well. Let's do that.